common/flatpak-dir.c

```diff
@@ -8080,6 +8267,168 @@ flatpak_dir_remote_fetch_summary (FlatpakDir   *self,
+          g_autoptr(GError) local_error = NULL;
+
+          /* Make sure the branch is up to date. */
+          if (!flatpak_dir_fetch_remote_repo_metadata (self, remote, state->summary, state->collection_id, cancellable, &local_error))
+            {
+              if (optional)
+                {
+                  /* This happens for instance in the case where a p2p remote is invalid (wrong signature)
+                     and we should just silently fail to update to it. */
+                  g_debug ("Failed to download optional metadata");
+                }
+              else
+                {
+                  g_propagate_error (error, g_steal_pointer (&local_error));
+                  return NULL;
+                }
+            }
+          else
+            {
+              /* Look up the commit containing the latest repository metadata. */
+              latest_rev = flatpak_dir_read_latest (self, remote, OSTREE_REPO_METADATA_REF,
+                                                    NULL, cancellable, error);
+              if (latest_rev == NULL)
+                return NULL;
+
+              if (!ostree_repo_load_commit (self->repo, latest_rev, &commit_v, NULL, error))
+                return NULL;
+
+              state->metadata = g_variant_get_child_value (commit_v, 0);
+            }
+#else  /* if !FLATPAK_ENABLE_P2P */
+          g_assert_not_reached ();
+#endif  /* !FLATPAK_ENABLE_P2P */
+        }
+    }
+
+  return g_steal_pointer (&state);
+}
+
+FlatpakRemoteState *
+flatpak_dir_get_remote_state (FlatpakDir   *self,
+                              const char   *remote,
+                              GCancellable *cancellable,
+                              GError      **error)
+{
+  return _flatpak_dir_get_remote_state (self, remote, FALSE, NULL, NULL, cancellable, error);
+}
+
+/* This is an alternative way to get the state where the summary is
+ * from elsewhere. It is mainly used by the system-helper where the
+ * summary is from the user-mode part which downloaded an update
+ *
+ * It will verify the summary if a signature is passed in, but not
+ * otherwise.
+ **/
+FlatpakRemoteState *
+flatpak_dir_get_remote_state_for_summary (FlatpakDir   *self,
+                                          const char   *remote,
+                                          GBytes       *opt_summary,
+                                          GBytes       *opt_summary_sig,
+                                          GCancellable *cancellable,
+                                          GError      **error)
+{
+  return _flatpak_dir_get_remote_state (self, remote, FALSE, opt_summary, opt_summary_sig, cancellable, error);
+}
+
+
+/* This is an alternative way to get the remote state that doesn't
+ * error out if the summary or metadata is not available in the p2p
+ * case. For example, we want to be able to update an app even when
+ * we can't talk to the main repo, but there is a local (p2p/sdcard)
+ * source for apps.
```

Comment:
In the offline P2P case, the `ostree-metadata` ref should always still be available. Please don’t treat it as optional. When `collection_id != NULL`, the `summary{,.sig}` files are optional, but the `ostree-metadata` ref is not.

Reply:
Well, in this particular code we essentially call ostree_repo_remote_fetch_summary and use that to find the ostree branch+commit for the metadata to pull. And the first fails if you are offline. This codepath is ok without metadata though.

However there are some problems here. Lemme file another issue about the P2P issues i've found while working on this stuff.